### ISSUE TITLE
Add iterators

### ISSUE THREAD

**Issue by Yamakaky** [BurntSushi/byteorder#58]
Like

```
for word in file.iter_u16<byteorder::LittleEndian>() {
    ...
}
```

**Comment by Yamakaky** [comment]
And the same for `Write` types

**Comment by BurntSushi** [comment]
This doubles the size of the trait unfortunately (given the current API), which is the only downside I can think of.

> And the same for Write types

Could you write out the type signatures for that? What are the semantics?

**Comment by Yamakaky** [comment]
hum, maybe not iterators, but at least a `write_all` thing.

**Comment by BurntSushi** [comment]
Could you please provide more details? I don't know what a "`write_all` thing" is. :-/ It'd be helpful to provide a full type signature and a short contract.

**Comment by Yamakaky** [comment]
https://gist.github.com/anonymous/6daf2ce56aea52d7615e1ddbd3e38ffc
Working on adding the little/bug endianness stuff

**Comment by Yamakaky** [comment]
And second version with endianess! https://gist.github.com/d0ec2e339ed0a87484519091bac8d740

**Comment by BurntSushi** [comment]
I'll just say this up front: I don't want this issue to turn into a combinatoric explosion of new methods, and I really don't want to add new traits. Adding iterators to `ReadBytesExt` *seems* like a reasonable addition though.

I'd recommend reading through #27.

**Comment by Yamakaky** [comment]
Well, my version has only on method while yours has 10 or so, so it's not really a combinatorial explosion.

**Comment by Yamakaky** [comment]
Last version with iterator: https://gist.github.com/9347c1f326ce8b5bcea0b0bbf6ec12b4

**Comment by BurntSushi** [comment]
Please see #27. We've already discussed that version of the API.

If you want to completely redesign the API and have new arguments not in #27, then open a new issue. Otherwise, I'm finding it incredibly hard to follow what it is you want or what problems you're trying to solve.

**Comment by Yamakaky** [comment]
I know, I started with the iterator and ended completely rewriting the API XD

**Comment by BurntSushi** [comment]
I'm going to close this because I'm not exactly sure how useful an iterator is here and in the current design at least, I believe it would require several additional methods, so I'm not sure it's worth it.

**Comment by ticki** [comment]
I'm sad this was closed, I really needed it for a project of mine. I propose adding a trait, `Encoding`, which is essentially an integer type which can be (de)serialized to/from bytes. This would make such an addition relatively simple.

**Comment by ticki** [comment]
#54

**Comment by Yamakaky** [comment]
For the record, I ended doing https://github.com/Yamakaky/dcpu/blob/master/src/byteorder.rs
